pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PartialUpsertTableRebalanceIntegrationTest.java

```diff
@@ -0,0 +1,458 @@
+  private static final String REALTIME_TABLE_NAME = TableNameBuilder.REALTIME.tableNameWithType(DEFAULT_TABLE_NAME);
+
+  // Segment 1 contains records of pk value 100000 (partition 0)
+  private static final String UPLOADED_SEGMENT_1 = "mytable_10027_19736_0 %";
+  // Segment 2 contains records of pk value 100001 (partition 1)
+  private static final String UPLOADED_SEGMENT_2 = "mytable_10072_19919_1 %";
+  // Segment 3 contains records of pk value 100002 (partition 1)
+  private static final String UPLOADED_SEGMENT_3 = "mytable_10158_19938_2 %";
+
+  private PinotHelixResourceManager _resourceManager;
+  private TableRebalancer _tableRebalancer;
+  private static List<File> _avroFiles;
+  private TableConfig _tableConfig;
+  private Schema _schema;
+
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+
+    startZk();
+    startController();
+    startBroker();
+    startServers(NUM_SERVERS);
+
+    // Start Kafka and push data into Kafka
+    startKafka();
+
+    _resourceManager = getControllerStarter().getHelixResourceManager();
+    _tableRebalancer = new TableRebalancer(_resourceManager.getHelixZkManager(), null);
+
+    createSchemaAndTable();
+  }
+
+  @Test
+  public void testRebalance()
+      throws Exception {
+    populateTables();
+
+    verifyIdealState(5, NUM_SERVERS);
+
+    // setup the rebalance config
+    Configuration rebalanceConfig = new BaseConfiguration();
+    rebalanceConfig.addProperty(RebalanceConfigConstants.DRY_RUN, false);
+    rebalanceConfig.addProperty(RebalanceConfigConstants.MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME, 0);
+    rebalanceConfig.addProperty(RebalanceConfigConstants.INCLUDE_CONSUMING, true);
+
+    // Add a new server
+    BaseServerStarter serverStarter1 = startOneServer(1234);
+
+    // Now we trigger a rebalance operation
+    TableConfig tableConfig = _resourceManager.getTableConfig(REALTIME_TABLE_NAME);
+    RebalanceResult rebalanceResult = _tableRebalancer.rebalance(tableConfig, rebalanceConfig);
+
+    // Check the number of replicas after rebalancing
+    int finalReplicas = _resourceManager.getServerInstancesForTable(getTableName(), TableType.REALTIME).size();
+
+    // Check that a replica has been added
+    assertEquals(finalReplicas, NUM_SERVERS + 1, "Rebalancing didn't correctly add the new server");
+
+    waitForRebalanceToComplete(rebalanceResult, 600_000L);
+    waitForAllDocsLoaded(600_000L);
+
+    verifySegmentAssignment(rebalanceResult.getSegmentAssignment(), 5, finalReplicas);
+
+    // Add a new server
+    BaseServerStarter serverStarter2 = startOneServer(4567);
+    rebalanceResult = _tableRebalancer.rebalance(tableConfig, rebalanceConfig);
+
+    // Check the number of replicas after rebalancing
+    finalReplicas = _resourceManager.getServerInstancesForTable(getTableName(), TableType.REALTIME).size();
+
+    // Check that a replica has been added
```

Comment:
I assume you meant a server has been added? as from the comment below, `rf=1`

Reply:
yes